tools/libxl/libxl_arm.c

```diff
@@ -786,29 +803,84 @@ static int copy_partial_fdt(libxl__gc *gc, void *fdt, void *pfdt)
     return 0;
 }
 
-static int copy_coprocs_nodes(libxl__gc *gc, void *fdt, void *pfdt,
-                              const libxl_domain_build_info *info)
-{
-    int i, rc;
-
-    LOG(DEBUG, "Copy coprocs nodes from the partial FDT");
+struct backlink {
+    int nodeoff;
+    int copied;
+    struct backlink *parent;
+};
 
-    for (i = 0; i < info->num_coprocs; i++) {
-        const libxl_device_coproc *coproc = &info->coprocs[i];
+static int copy_ancestor(libxl__gc *gc, void *fdt, void *pfdt,
+                         struct backlink *nodelink)
+{
+    int ret = 0;
+
+    if (nodelink && nodelink->copied == 0)
+    {
+        copy_ancestor(gc, fdt, pfdt, nodelink->parent);
+        LOG(DEBUG, "Begin node %s", fdt_get_name(pfdt, nodelink->nodeoff,
+            NULL));
+        ret = fdt_begin_node(fdt, fdt_get_name(pfdt, nodelink->nodeoff, NULL));
+        if (ret) return ret;
+        
+        ret = copy_properties(gc, fdt, pfdt, nodelink->nodeoff);
+        if (ret) return ret;
+        nodelink->copied = 1;
+    }
+    return ret;
+}
 
-        if (!coproc->path) {
-            LOG(ERROR, "Bad coproc node for the partial FDT");
-            continue;
+static int descent_for_vcoproc(libxl__gc *gc, void *fdt, void *pfdt,
+                               struct backlink *nodelink)
+{
+    int nodeoff;
+    int ret = 0;
+    struct backlink node;
+
+    LOG(DEBUG, "process node %s", fdt_get_name(pfdt, nodelink->nodeoff, NULL));
+    if ( fdt_getprop(pfdt, nodelink->nodeoff, "xen,vcoproc", NULL) )
+    {
+        LOG(DEBUG, "\"xen,vcoproc\" is found, copy ancestors");
+        ret = copy_ancestor(gc, fdt, pfdt, nodelink->parent);
+        if (ret) return ret;
+        LOG(DEBUG, "\"xen,vcoproc\" is found, copy node %s", fdt_get_name(pfdt,
+            nodelink->nodeoff, NULL));
```

Comment:
You call fdt_get_name(pfdt, nodelink->nodeoff, NULL) twice. Maybe introduce a variable?